File: server/src/api/chapters.rs
//! `/api/chapters` — port of `Api/Chapters.cs` (ADR 0033, S5 #127). Flipped
//! in `docker/nginx.conf` as a block with `titles`/`progress`/`queue` now
//! that S7 (#129) landed (see `crate::titles`'s module doc).
//!
//! `VerifyDownloadsAsync` (pruning `downloaded=1` rows whose file vanished)
//! is a no-op TODO stub on the .NET side too — nothing to port yet.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use serde::Serialize;

use crate::auth::Claims;
use crate::chapters;
use crate::error::{AppError, AppResult};
use crate::state::AppState;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/title/{title_id}", get(list))
        .route("/{id}", get(get_chapter))
        .route("/{id}/text", get(get_chapter_text))
        .route("/{id}/download", axum::routing::post(queue_download))
}

#[derive(Serialize)]
struct ChapterDto {
    id: String,
    title_id: String,
    title: Option<String>,
    number: f64,
    volume: Option<f64>,
    local_path: Option<String>,
    page_count: i64,
    downloaded: bool,
    has_sources: bool,
    chapter_format: String,
    created_at: String,
}

impl From<chapters::ChapterRow> for ChapterDto {
    fn from(c: chapters::ChapterRow) -> Self {
        Self {
            id: c.id,
            title_id: c.title_id,
            title: c.title,
            number: c.number,
            volume: c.volume,
            local_path: c.local_path,
            page_count: c.page_count,
            downloaded: c.downloaded,
            has_sources: c.has_sources,
            chapter_format: c.chapter_format,
            created_at: c.created_at.replacen(' ', "T", 1),
        }
    }
}

async fn list(
    claims: Claims,
    State(state): State<AppState>,
    Path(title_id): Path<String>,
) -> AppResult<Json<Vec<ChapterDto>>> {
    let rows = chapters::list_chapters(&state.db, &title_id, claims.allow_explicit).await?;
    Ok(Json(rows.into_iter().map(ChapterDto::from).collect()))
}

async fn get_chapter(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<Json<ChapterDto>> {
    let c = chapters::get_chapter(&state.db, &id, claims.allow_explicit)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(ChapterDto::from(c)))
}

#[derive(Serialize)]
struct ChapterTextDto {
    content: String,
}

async fn get_chapter_text(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<Json<ChapterTextDto>> {
    let info = chapters::chapter_text_info(&state.db, &id, claims.allow_explicit)
        .await?
        .ok_or(AppError::NotFound)?;

    if info.chapter_format != "text" {
        return Err(AppError::BadRequest("not a text chapter".into()));
    }
    if !info.downloaded {
        return Err(AppError::NotFound);
    }
    let Some(path) = info.local_path else {
        return Err(AppError::NotFound);
    };

    let content = tokio::fs::read_to_string(&path)
        .await
        .map_err(|_| AppError::NotFound)?;
    Ok(Json(ChapterTextDto { content }))
}

async fn queue_download(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<StatusCode> {
    let candidate = chapters::queue_candidate(&state.db, &id, claims.allow_explicit)
        .await?
        .ok_or(AppError::NotFound)?;

    chapters::queue_download(
        &state.db,
        &candidate.id,
        &candidate.manga_title,
        candidate.number,
        &claims.user_id,
    )
    .await?;

    Ok(StatusCode::ACCEPTED)
}

File: server/Api/Chapters.cs
using System.Linq.Expressions;
using System.Security.Claims;
using ArrghServer.Data;
using ArrghServer.Data.Models;
using Microsoft.EntityFrameworkCore;

namespace ArrghServer.Api;

public static class Chapters
{
    public static RouteGroupBuilder MapChaptersRoutes(this RouteGroupBuilder group)
    {
        group.MapGet("/title/{titleId}", ListChapters).RequireAuthorization().WithSummary("List chapters for a title");
        group.MapGet("/{id}", GetChapter).RequireAuthorization().WithSummary("Get a single chapter with page list");
        group.MapGet("/{id}/text", GetChapterText).RequireAuthorization().WithSummary("Get chapter text content (novels)");
        group.MapPost("/{id}/download", QueueDownload).RequireAuthorization().WithSummary("Queue a chapter for download");
        return group;
    }

    // -------------------------------------------------------------------------

    static async Task<IResult> ListChapters(string titleId, ClaimsPrincipal principal, AppDbContext db)
    {
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";

        var chapters = await db.Chapters
            .Where(c => c.TitleId == titleId
                     && (c.Title.IsExplicit == false || allowExplicit))
            .OrderBy(c => c.Number)
            .Select(ChapterProjection)
            .ToListAsync();

        _ = Task.Run(() => VerifyDownloadsAsync(db, titleId));

        return Results.Ok(chapters);
    }

    static async Task<IResult> GetChapter(string id, ClaimsPrincipal principal, AppDbContext db)
    {
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";

        var chapter = await db.Chapters
            .Where(c => c.Id == id && (c.Title.IsExplicit == false || allowExplicit))
            .Select(ChapterProjection)
            .FirstOrDefaultAsync();

        if (chapter is null) return Results.NotFound();
        return Results.Ok(chapter);
    }

    static async Task<IResult> GetChapterText(string id, ClaimsPrincipal principal, AppDbContext db)
    {
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";

        var row = await db.Chapters
            .Where(c => c.Id == id && (c.Title.IsExplicit == false || allowExplicit))
            .Select(c => new { c.LocalPath, c.Downloaded, c.ChapterFormat })
            .FirstOrDefaultAsync();

        if (row is null) return Results.NotFound();
        if (row.ChapterFormat != "text") return Results.BadRequest();
        if (!row.Downloaded) return Results.NotFound();
        if (row.LocalPath is null) return Results.NotFound();

        try
        {
            var content = await File.ReadAllTextAsync(row.LocalPath);
            return Results.Ok(new { content });
        }
        catch
        {
            return Results.NotFound();
        }
    }

    static async Task<IResult> QueueDownload(string id, ClaimsPrincipal principal, AppDbContext db)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";

        var candidate = await db.Chapters
            .Where(c => c.Id == id
                     && !c.Downloaded
                     && c.ChapterSources.Any()
                     && (c.Title.IsExplicit == false || allowExplicit))
            .Select(c => new { c.Id, c.Number, MangaTitle = c.Title.TitleName })
            .FirstOrDefaultAsync();

        if (candidate is null) return Results.NotFound();

        var queueId = Guid.NewGuid().ToString();
        var now = DateTime.UtcNow;

        // RAW SQL — intentional: SQLite upsert with conditional WHERE on status.
        // No EF equivalent for ON CONFLICT ... DO UPDATE ... WHERE.
        // If this becomes hard to maintain, consider: read existing row + conditional SaveChanges.
        await db.Database.ExecuteSqlRawAsync(@"
            INSERT INTO download_queue
                (id, chapter_id, manga_title, chapter_num, status, pages_downloaded, pages_total, created_at, updated_at, queued_by)
            VALUES ({0}, {1}, {2}, {3}, 'pending', 0, 0, {4}, {5}, {6})
            ON CONFLICT(chapter_id) DO UPDATE SET
                status = 'pending',
                error = NULL,
                queued_by = excluded.queued_by,
                updated_at = excluded.updated_at
            WHERE download_queue.status IN ('error', 'cancelled')",
            queueId, candidate.Id, candidate.MangaTitle, candidate.Number, now, now, userId);

        return Results.StatusCode(202);
    }

    // -------------------------------------------------------------------------

    // Expression tree — EF translates HasSources = c.ChapterSources.Any() to a SQL EXISTS subquery.
    // Must be Expression<Func<...>>, NOT a plain method (plain method → client-side eval → empty navigations).
    static Expression<Func<Chapter, ChapterDto>> ChapterProjection =>
        c => new ChapterDto
        {
            Id = c.Id,
            TitleId = c.TitleId,
            Title = c.ChapterTitle,
            Number = c.Number,
            Volume = c.Volume,
            LocalPath = c.LocalPath,
            PageCount = c.PageCount,
            Downloaded = c.Downloaded,
            HasSources = c.ChapterSources.Any(),
            ChapterFormat = c.ChapterFormat,
            CreatedAt = c.CreatedAt,
        };

    static async Task VerifyDownloadsAsync(AppDbContext db, string titleId)
    {
        // TODO: port from Rust indexer/local.rs verify_title_downloads
        // Resets downloaded=1 rows whose local_path file no longer exists on disk
        await Task.CompletedTask;
    }
}

// ── DTOs ──────────────────────────────────────────────────────────────────────

public class ChapterDto
{
    public string Id { get; set; } = null!;
    public string TitleId { get; set; } = null!;
    public string? Title { get; set; }
    public double Number { get; set; }
    public double? Volume { get; set; }
    public string? LocalPath { get; set; }
    public int PageCount { get; set; }
    public bool Downloaded { get; set; }
    public bool HasSources { get; set; }
    public string ChapterFormat { get; set; } = null!;
    public DateTime CreatedAt { get; set; }
}
